display_list/geometry/dl_region.cc

```diff
@@ -8,31 +8,527 @@
+        start_x = r->left();
+        end_x = r->right();
+      } else if (end_x < r->right()) {
+        end_x = r->right();
+      }
+      if (end_y > r->bottom()) {
+        end_y = r->bottom();
+      }
+    }
+    working_spans.emplace_back(start_x, end_x);
+
+    // end_y must not pass by the top of the next input rect
+    if (next_rect < count && end_y > rects[next_rect]->top()) {
+      end_y = rects[next_rect]->top();
+    }
+
+    // If all of the rules above work out, we should never collapse the
+    // current range of Y coordinates to empty
+    FML_DCHECK(end_y > cur_y);
+
+    if (!lines_.empty() && lines_.back().bottom == cur_y &&
+        spansEqual(lines_.back(), working_spans.data(),
+                   working_spans.data() + working_spans.size())) {
+      lines_.back().bottom = end_y;
+    } else {
+#ifdef DlRegion_DO_STATS
+      span_count += working_spans.size();
+      line_count++;
+#endif
+      lines_.push_back(makeLine(cur_y, end_y, working_spans));
+    }
+    cur_y = end_y;
+  }
+
+#ifdef DlRegion_DO_STATS
+  double span_avg = ((double)span_count) / line_count;
+  double active_avg = ((double)active_rect_count) / pass_count;
+  FML_LOG(ERROR) << lines_.size() << " lines for " << count
+                 << " input rects, avg " << span_avg
+                 << " spans per line and avg " << active_avg
+                 << " active rects per loop";
+#endif
+}
+
+void DlRegion::appendLine(int32_t top,
+                          int32_t bottom,
+                          const Span* begin,
+                          const Span* end) {
+  if (lines_.empty()) {
+    lines_.push_back(makeLine(top, bottom, begin, end));
+  } else {
+    if (lines_.back().bottom == top && spansEqual(lines_.back(), begin, end)) {
+      lines_.back().bottom = bottom;
+    } else {
+      lines_.push_back(makeLine(top, bottom, begin, end));
+    }
+  }
+}
+
+DlRegion DlRegion::MakeUnion(const DlRegion& a, const DlRegion& b) {
+  DlRegion res;
+
+  res.bounds_ = a.bounds_;
+  res.bounds_.join(b.bounds_);
+
+  res.span_buffer_.reserve(a.span_buffer_.capacity() +
+                           b.span_buffer_.capacity());
+
+  auto& lines = res.lines_;
+  lines.reserve(a.lines_.size() + b.lines_.size());
+
+  auto a_lines = a.lines_;
+  auto b_lines = b.lines_;
+  auto a_it = a_lines.begin();
+  auto b_it = b_lines.begin();
+
+  auto& a_buffer = a.span_buffer_;
+  auto& b_buffer = b.span_buffer_;
+
+  std::vector<Span> tmp;
+
+  while (a_it != a_lines.end() && b_it != b_lines.end()) {
+    if (a_it->bottom <= b_it->top) {
+      res.appendLine(a_it->top, a_it->bottom, a_buffer, a_it->chunk_handle);
+      ++a_it;
+    } else if (b_it->bottom <= a_it->top) {
+      res.appendLine(b_it->top, b_it->bottom, b_buffer, b_it->chunk_handle);
+      ++b_it;
+    } else {
+      if (a_it->top < b_it->top) {
+        res.appendLine(a_it->top, b_it->top, a_buffer, a_it->chunk_handle);
+        a_it->top = b_it->top;
```

Comment:
Does this modify data in the a region's lines?

(Ditto for the same line in the b variant below.)

Reply:
Perhaps have a y value that you use for tracking where you're at without modifying the iterated lines?

Reply:
Hmmm, the component regions are declared const, shouldn't that result in const iterators?

Reply:
Maybe we need to declare the lines_ and the data in their SpanVec's to all be const vectors of const data?

Reply:
Done.

Reply:
I think the caller would be surprised if the component regions were modified, especially if it was a long-lived object that they will be reusing. Even if it were somewhat temporary they might want to ask one of the component regions a question without realizing it had been modified.